x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java

```diff
@@ -106,15 +138,16 @@ private float[] extractScoresFromRankedDocs(List<RankedDocsResults.RankedDoc> ra
         return scores;
     }
 
-    /**
-     * Sorts documents by score descending and discards those with a score less than minScore.
-     * @param originalDocs documents to process
-     */
-    @Override
-    protected RankFeatureDoc[] preprocess(RankFeatureDoc[] originalDocs) {
-        return Arrays.stream(originalDocs)
-            .filter(doc -> minScore == null || doc.score >= minScore)
-            .sorted(Comparator.comparing((RankFeatureDoc doc) -> doc.score).reversed())
-            .toArray(RankFeatureDoc[]::new);
+    private Integer getTopNFromInferenceEndpointTaskSettings() {
+        GetInferenceModelAction.Request request = new GetInferenceModelAction.Request(inferenceId, TaskType.RERANK);
+        ActionFuture<GetInferenceModelAction.Response> response = client.execute(GetInferenceModelAction.INSTANCE, request);
+        ModelConfigurations modelConfigurations = response.actionGet().getEndpoints().get(0);
+
+        return modelConfigurations.getService().equals(CohereService.NAME)
+            && modelConfigurations.getTaskType().equals(TaskType.RERANK)
+            && modelConfigurations.getTaskSettings() instanceof CohereRerankTaskSettings
+                ? ((CohereRerankTaskSettings) modelConfigurations.getTaskSettings()).getTopNDocumentsOnly()
+                : null;
```

Comment:
Lets not use an "ActionFuture" this will take a thread and block on it. We should instead keep it all asynchronous.